### PR TITLE
[IOTDB-5310] Added removing status to region migration

### PR DESCRIPTION
**Pull request by Caideyipi**
## IOTDB-5310
image
image
image

<hr>

This PR has:
- [ ] been self-reviewed.
    - [ ] concurrent read
    - [ ] concurrent write
    - [ ] concurrent read and write 
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. 
- [ ] added or updated version, __license__, or notice information
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious 
  for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold 
  for code coverage.
- [ ] added integration tests.
- [ ] been tested in a test IoTDB cluster.

<hr>

### CODE REVIEW

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/partition/heartbeat/RegionCache.java [hunk 1]
```diff
@@ -31,6 +31,8 @@ public class RegionCache {
 
   private final List<RegionHeartbeatSample> slidingWindow;
 
+  private boolean is_migrating = false;
```
Comment (Beyyes) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/partition/heartbeat/RegionCache.java:34]: isMigrating

File: confignode/src/main/java/org/apache/iotdb/confignode/procedure/env/DataNodeRemoveHandler.java [hunk 2]
```diff
@@ -75,6 +75,16 @@ public static String getIdWithRpcEndpoint(TDataNodeLocation location) {
         location.getDataNodeId(), location.getClientRpcEndPoint());
   }
 
+  public void markRegionAsMigrating(
+      TConsensusGroupId consensusGroupId, TDataNodeLocation originalDataNode) {
+    configManager
+        .getPartitionManager()
+        .getRegionGroupCacheMap()
+        .get(consensusGroupId)
+        .getRegionCache(originalDataNode.getDataNodeId())
```
Comment (Beyyes) [inline confignode/src/main/java/org/apache/iotdb/confignode/procedure/env/DataNodeRemoveHandler.java:84]: add `get(consensusGroupId)` and `.getRegionCache` npe judgement

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/partition/heartbeat/RegionCache.java [hunk 3]
```diff
@@ -70,4 +75,8 @@ public RegionStatistics getRegionStatistics() {
   private RegionHeartbeatSample getLastSample() {
     return slidingWindow.get(slidingWindow.size() - 1);
   }
+
+  public void setIsMigrating(boolean isMigrating) {
```
Comment (Beyyes) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/partition/heartbeat/RegionCache.java:79]: setMigrating is better?

File: confignode/src/main/java/org/apache/iotdb/confignode/procedure/impl/statemachine/RegionMigrateProcedure.java [hunk 4]
```diff
@@ -87,6 +87,7 @@ protected Flow executeFromState(ConfigNodeProcedureEnv env, RegionTransitionStat
     try {
       switch (state) {
         case REGION_MIGRATE_PREPARE:
+          handler.markRegionAsMigrating(consensusGroupId, originalDataNode);
```
Comment (CRZbulabula) [inline confignode/src/main/java/org/apache/iotdb/confignode/procedure/impl/statemachine/RegionMigrateProcedure.java:90]: This is a effective but not safe way to update RegionStatus. The migrating status will be lost if the ConfigNodeGroup switch the ConfigNode-leader during Region migration process. A better way to satisfy this feature is sending a RPC request to record this Removing status on DataNode. Therefore this Removing status will be sent to ConfigNode-leader in Cluster heartbeat. You can reference how ConfigNode currently identifies a Removing DataNode.

**Review by CRZbulabula — CHANGES_REQUESTED** [review]
PTAL

**Comment by Beyyes** [comment]
Follow up work: https://issues.apache.org/jira/browse/IOTDB-5419

**Pull request closed without merging**